src/NodaTime/LocalDate.cs

```diff
@@ -627,7 +627,9 @@ public ZonedDateTime AtStartOfDayInZone(DateTimeZone zone)
         public LocalDate WithCalendar(CalendarSystem calendar)
         {
             Preconditions.CheckNotNull(calendar, nameof(calendar));
-            return new LocalDate(DaysSinceEpoch, calendar);
+            return ReferenceEquals(calendar, Calendar)
```

Comment:
Rather than use the Calendar property, which has to look up the CalendarSystem via its Ordinal, I'd just compare the ordinals: `calendar.Ordinal == CalendarOrdinal`.

Reply:
Ahh, i missed that subtlety, I just assumed the calendar was stored as a pointer. That's for catching that, this makes the performance improvement even more significant.

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3880/23H2/2023Update/SunValley3)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 8.0.303
  [Host]     : .NET 8.0.7 (8.0.724.31311), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.7 (8.0.724.31311), X64 RyuJIT AVX2


| Method                          | Mean      | Error     | StdDev    | Ratio |
|-------------------------------- |----------:|----------:|----------:|------:|
| WithCalendar_Unchanged_New      | 0.0836 ns | 0.0096 ns | 0.0085 ns |  0.01 |
| WithCalendar_Unchanged_Previous | 7.1869 ns | 0.1479 ns | 0.1383 ns |  1.00 |